user/syscalls.S

```diff
@@ -0,0 +1,67 @@
+#include <mips/asm.h>
+#include <errno.h>
+
+    .set noreorder
+    .set noat
+
+/* This implementation uses newlib-specific errno processing. */
+#define DO_SYSCALL(num)  \
+    addiu $sp, $sp, -16; \
+    sw $ra, 12($sp);     \
+    li $v0, num;         \
+    syscall;             \
+    bgez $v0, 1f;        \
+    nop;                 \
+    sw $v0, 8($sp);      \
+    jal __errno;         \
+    nop;                 \
+    lw $v1, 8($sp);      \
+    negu $v1, $v1;       \
+    sw $v1, 0($v0);      \
+    li $v0, -1;          \
+1:  lw $ra, 12($sp);     \
+    addiu $sp, $sp, 16;  \
+    jr $ra;              \
+    nop;
+
+#define DO_NO_SYSCALL() \
+    addiu $sp, $sp, -8; \
+    sw $ra, 4($sp);     \
+    jal __errno;        \
+    nop;                \
+    li $v1, ENOSYS;     \
+    sw $v1, 0($v0);     \
+    li $v0, -1;         \
+    lw $ra, 4($sp);     \
+    addiu $sp, $sp, 8;  \
+    jr $ra;             \
+    nop;
+
+#define SYSCALL(name, num) \
+    LEAF(name)      \
+    DO_SYSCALL(num) \
+    END(name)
+#define NO_SYSCALL(name) \
```

Comment:
I really can't figure out how to use `ALEAF`. It looks to me that it may only be used to declare a procedure and it's *single* alias, right?